### PR TITLE
fixes #18489: provide support of array attributes in multipart/form-data request for angular typescript client

### LINKED ISSUE #18489: [BUG][Typescript][Angular] Wrong generated code for array of items using FormData (multipart/form-data)
#### Bug Report Checklist

- [ x] Have you provided a full/minimal spec to reproduce the issue?
- [ x] Have you validated the input using an OpenAPI validator ([example](https://apidevtools.org/swagger-parser/online/))?
- [x ] Have you [tested with the latest master](https://github.com/OpenAPITools/openapi-generator/wiki/FAQ#how-to-test-with-the-latest-master-of-openapi-generator) to confirm the issue still exists?
- [ x] Have you searched for related issues/PRs?
- [ x] What's the actual output vs expected output?
- [x] [Optional] Sponsorship to speed up the bug fix or feature request ([example](https://github.com/OpenAPITools/openapi-generator/issues/6178))

##### Description

I use open api spec 3 to define api to define two attributes : 

* mediasToAdd which is an array of files to upload

* mediasToDelete which is an array of files uuids to delete

-------

But Angular Typescript generators generates a code that does not build properly theses array attributes.

This is the generated code : 

Capture d’écran 2024-04-24 à 18 05 11

But it should be : 

Capture d’écran 2024-04-24 à 18 05 31

-> If I use the generated method to upsert my medias, I can observe this in the chrome developers network section

Capture d’écran 2024-04-24 à 18 13 54

There are multiple formData attributes with the same name which are sent,

the consequence is that on API side theses values are received as simple attributes instead as arrays.

-------

Following the previous screenshot the API will receive : 

* mediasToAdd as a single binary (latest value provided)

* mediasToDelete as a single string (latest value provided)

##### openapi-generator version

I use exactly @openapitools/openapi-generator-cli@1.0.18-5.0.0-beta2

##### OpenAPI declaration file content or url

Capture d’écran 2024-04-24 à 17 54 45

##### Suggest a fix

The problem is located in this file => https://github.com/OpenAPITools/openapi-generator/blob/master/modules/openapi-generator/src/main/resources/typescript-angular/api.service.mustache

The attributes are appended as simple attributes instead as array : 

```yaml
        {{#isArray}}
        if ({{paramName}}) {
        {{#isCollectionFormatMulti}}
            {{paramName}}.forEach((element) => {
                localVarFormParams = localVarFormParams.append('{{baseName}}', <any>element) as any || localVarFormParams;
            })
        {{/isCollectionFormatMulti}}
        {{^isCollectionFormatMulti}}
            if (localVarUseForm) {
                {{paramName}}.forEach((element) => {
                    localVarFormParams = localVarFormParams.append('{{baseName}}', <any>element) as any || localVarFormParams;
            })
            } else {
                localVarFormParams = localVarFormParams.append('{{baseName}}', [...{{paramName}}].join(COLLECTION_FORMATS['{{collectionFormat}}'])) as any || localVarFormParams;
            }
        {{/isCollectionFormatMulti}}
        }
        {{/isArray}}
```

But it should be this 

```yaml
        {{#isArray}}
        if ({{paramName}}) {
        {{#isCollectionFormatMulti}}
            {{paramName}}.forEach((element) => {
                localVarFormParams = localVarFormParams.append('{{baseName}}[]', <any>element) as any || localVarFormParams;
            })
        {{/isCollectionFormatMulti}}
        {{^isCollectionFormatMulti}}
            if (localVarUseForm) {
                {{paramName}}.forEach((element) => {
                    localVarFormParams = localVarFormParams.append('{{baseName}}[]', <any>element) as any || localVarFormParams;
            })
            } else {
                localVarFormParams = localVarFormParams.append('{{baseName}}[]', [...{{paramName}}].join(COLLECTION_FORMATS['{{collectionFormat}}'])) as any || localVarFormParams;
            }
        {{/isCollectionFormatMulti}}
        }
        {{/isArray}}
```

### PR DESCRIPTION
**Pull request by jeandonaldroselin**
If first declared the issue there => #18489 

### PR checklist
 
- [x] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [x] Pull Request title clearly describes the work in the pull request and Pull Request description provides details about how to validate the work. Missing information here may result in delayed response from the community.
- [x] Run the following to [build the project](https://github.com/OpenAPITools/openapi-generator#14---build-projects) and update samples:
  ```
  ./mvnw clean package 
  ./bin/generate-samples.sh ./bin/configs/*.yaml
  ./bin/utils/export_docs_generators.sh
  ``` 
  (For Windows users, please run the script in [Git BASH](https://gitforwindows.org/))
  Commit all changed files. 
  This is important, as CI jobs will verify _all_ generator outputs of your HEAD commit as it would merge with master. 
  These must match the expectations made by your contribution. 
  You may regenerate an individual generator by passing the relevant config(s) as an argument to the script, for example `./bin/generate-samples.sh bin/configs/java*`. 
  IMPORTANT: Do **NOT** purge/delete any folders/files (e.g. tests) when regenerating the samples as manually written tests may be removed.
- [x] File the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master` (upcoming 7.1.0 minor release - breaking changes with fallbacks), `8.0.x` (breaking changes without fallbacks)
- [x] If your PR is targeting a particular programming language, @mention the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) members, so they are more likely to review the pull request.

Please review my PR : 
@TiFu @taxpon @sebastianhaas @kenisteward @Vrolijkx @macjohnny @topce @akehir @petejohansonxo @amakhrov @davidgamero @mkusaka

### CODE REVIEW

**Review by macjohnny — CHANGES_REQUESTED** [review]
I dont think this change can be done without introducing an option. The square brackets are a convention for languages like PHP, but not a standard, see eg https://stackoverflow.com/a/29371999

**Comment by jeandonaldroselin** [comment]
Hello @macjohnny,

Thank you for your suick reply, I did not see github notifications.

I added an option named useSquareBracketsInArrayNames.
Setting this property to true will add brackets to array attribute names.

**Comment by macjohnny** [comment]
please re-generate the docs

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/TypeScriptAngularClientCodegen.java [hunk 1]
```diff
@@ -138,6 +139,7 @@ public TypeScriptAngularClientCodegen() {
         this.cliOptions.add(new CliOption(FILE_NAMING, "Naming convention for the output files: 'camelCase', 'kebab-case'.").defaultValue(this.fileNaming));
         this.cliOptions.add(new CliOption(STRING_ENUMS, STRING_ENUMS_DESC).defaultValue(String.valueOf(this.stringEnums)));
         this.cliOptions.add(new CliOption(QUERY_PARAM_OBJECT_FORMAT, "The format for query param objects: 'dot', 'json', 'key'.").defaultValue(this.queryParamObjectFormat.name()));
+        this.cliOptions.add(CliOption.newBoolean(USE_SQUARE_BRACKETS_IN_ARRAY_NAMES, "Setting this property to true will add brackets to array attribute names.", false));
```
Comment (macjohnny) [inline modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/TypeScriptAngularClientCodegen.java:142]:
```suggestion
        this.cliOptions.add(CliOption.newBoolean(USE_SQUARE_BRACKETS_IN_ARRAY_NAMES, "Setting this property to true will add brackets to array attribute names, e.g. my_values[].", false));
```

Comment (jeandonaldroselin) [reply modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/TypeScriptAngularClientCodegen.java]: > 

Hello @macjohnny,

I updated the description as you suggested and the doc has been updated.
Please have a look.

Than you in advance :)

**Review by macjohnny — APPROVED** [review]

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/TypeScriptAngularClientCodegen.java (hunk 1 above)
Comment (jeandonaldroselin) [reply modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/TypeScriptAngularClientCodegen.java]: Hello @macjohnny,

Thank you for the review, I just pushed an update for TypeScriptAngularClientOptionsTest file
to fix a broken test case. The build is now passing, please can you review the PR again ?

Thank you in advance.

Comment (macjohnny) [reply modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/TypeScriptAngularClientCodegen.java]: Thanks, if the CI now passes, we can merge

**Comment by jeandonaldroselin** [comment]
Hello @macjohnny,

Thank you for the time you took to review 🙌🏾
I hope this fix will we usefull for a maximum of teams.
See you ++

**Pull request merged**